File: tp4/writer_reader/main.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>

#define MAX_BUF 1024

char *reader(int fd)
{
	int		err;
	char	buf[MAX_BUF];

	do {
		if ((err = read(fd, buf, MAX_BUF)) == 0) break;
		
		printf("%s \n", buf);
        	fflush(stdout);

	}while (strcmp(buf,"end"));

	return buf;
}

int writer(int	fd,char *buf)
{
	int i,len;
	write(fd, buf, len);
	write(fd, "end\n", sizeof("end"));		

	return 0;
}


int main()
{
	char    *cToPython = "cToPython";
	char    *pythonToC = "pythonToC";
	int	fd1,fd2;
	char a;

	char	buf[MAX_BUF];
	
	/* create the FIFO (named pipe) */
	mkfifo(cToPython, 0777);
	mkfifo(pythonToC, 0777);

	printf("open \n");
	fd1 = open(cToPython, O_RDWR|O_NONBLOCK);
	fd2 = open(pythonToC, O_RDWR|O_NONBLOCK);

	dup2(STDIN_FILENO, fd1);
	dup2(fd2, STDOUT_FILENO);

	writer(fd1,"toto");

	scanf("%s",&a);

	reader(fd2);

	/* remove the FIFO */
	close(fd1);
	close(fd2);
	unlink(cToPython);
	unlink(pythonToC);

return 0;
}

File: tp4/writer_reader/main.py
#!/usr/bin/env python
import os, time


def reader(pipe_in):
	
	string = ""

	string = pipe_in.readline()
	print '%s' % string,



def writer(pipe_out, input_str):

	pipe_out.write(input_str)
	pipe_out.flush()
	time.sleep(1)
	pipe_out.write("end\n")


def main():
	
	if not os.path.exists('pythonToC'):
	    os.mkfifo('pythonToC')

	if not os.path.exists('cToPython'):
	    os.mkfifo('cToPython')


	pythonToC = open('pythonToC','w')
	cToPython = open('cToPython','r')

	reader(cToPython)
	#time.sleep(1)
	writer(pythonToC, "received")
	#writer(pythonToC, "end")
	print 'send'


main()
